File: erl/normal_cdf_inverse.erl
-module(normal_cdf_inverse).
-export([normal_cdf_inverse/1]).

rational_approximation(T) ->
    % Abramowitz and Stegun formula 26.2.23.
    % The absolute value of the error should be less than 4.5 e-4.
    C0 = 2.515517, C1 = 0.802853, C2 = 0.010328,
    D0 = 1.432788, D1 = 0.189269, D2 = 0.001308,
    Numerator = (C2*T + C1)*T + C0,
    Denominator = ((D2*T + D1)*T + D0)*T + 1.0,
    T - Numerator / Denominator.

normal_cdf_inverse(P) ->
    % See article above for explanation of this section.
    if
        P < 0.5 ->
            % F^-1(p) = - G^-1(p)
            -rational_approximation( math:sqrt(-2.0*math:log(P)) );
        true ->
            % F^-1(p) = G^-1(1-p)
            rational_approximation( math:sqrt(-2.0*math:log(1.0-P)) )
    end.

File: pl/normal_cdf_inverse.pl
sub rational_approximation {
    my $t = $_[0];

    # Abramowitz and Stegun formula 26.2.23.
    # The absolute value of the error should be less than 4.5 e-4.
    my @c = (2.515517, 0.802853, 0.010328);
    my @d = (1.432788, 0.189269, 0.001308);
    my $numerator = ($c[2]*$t + $c[1])*$t + $c[0];
    my $denominator = (($d[2]*$t + $d[1])*$t + $d[0])*$t + 1.0;
    return $t - $numerator / $denominator;
}


sub normal_cdf_inverse {
    my $p = $_[0];

    # See article above for explanation of this section.
    if ($p < 0.5) {
        # F^-1(p) = - G^-1(p)
        return -rational_approximation( sqrt(-2.0*log($p)) );
    } else {
        # F^-1(p) = G^-1(1-p)
        return rational_approximation( sqrt(-2.0*log(1.0-$p)) );
    }
}

1;
